# 4.8 Tasks Grouping & Aggregating Variables

# Step 1 Create a new notebook, import libraries and 'ords_prods_merge' dataframe

In [1]:
# Importing libraries
import pandas as pd
import numpy as np
import os

In [ ]:
# Importing 'ords_prods_merged' dataframe
# path = r'D:\Rich\Bus Analyst\29-06-2023 Instacart Basket Analysis'
# 'ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged.pkl'))'

In [2]:
path = r'D:\Rich\Bus Analyst\29-06-2023 Instacart Basket Analysis'

In [3]:
path

'D:\\Rich\\Bus Analyst\\29-06-2023 Instacart Basket Analysis'

In [4]:
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged.pkl'))

In [ ]:
# Check consistency of 'ords_prods_merge' dataset

In [5]:
ords_prods_merge.head(10)

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   
5   3367565        1    prior             6          2                  7   
6    550135        1    prior             7          1                  9   
7   3108588        1    prior             8          1                 14   
8   2295261        1    prior             9          1                 16   
9   2550362        1    prior            10          4                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered _merge  \
0                     NaN         196                  1          0   both   
1                    15.0         196                  1          1   both   
2                    21.0         196                  1          1   both   
3                    29.0         196                  1          1   both   
4                    28.0         196                  1          1   both   
5                    19.0         196                  1          1   both   
6                    20.0         196                  1          1   both   
7                    14.0         196                  2          1   both   
8                     0.0         196                  4          1   both   
9                    30.0         196                  1          1   both   

  product_name  aisle_id  department_id  prices    price_range_loc  \
0         Soda        77              7     9.0  Mid-range product   
1         Soda        77              7     9.0  Mid-range product   
2         Soda        77              7     9.0  Mid-range product   
3         Soda        77              7     9.0  Mid-range product   
4         Soda        77              7     9.0  Mid-range product   
5         Soda        77              7     9.0  Mid-range product   
6         Soda        77              7     9.0  Mid-range product   
7         Soda        77              7     9.0  Mid-range product   
8         Soda        77              7     9.0  Mid-range product   
9         Soda        77              7     9.0  Mid-range product   

   orders_day_of_the_week  Busiest_days busiest_period_of_day  
0                       2  Regular busy        Average orders  
1                       3  Slowest days        Average orders  
2                       3  Slowest days        Average orders  
3                       4  Slowest days        Average orders  
4                       4  Slowest days        Average orders  
5                       2  Regular busy        Average orders  
6                       1  Busiest days        Average orders  
7                       1  Busiest days        Average orders  
8                       1  Busiest days        Average orders  
9                       4  Slowest days        Average orders

In [6]:
# Need to drop columns 'eval_set' and '_merge'

ords_prods_merge = ords_prods_merge.drop(columns = ['eval_set', '_merge'])

In [7]:
ords_prods_merge.head(10)

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0   2539329        1             1          2                  8   
1   2398795        1             2          3                  7   
2    473747        1             3          3                 12   
3   2254736        1             4          4                  7   
4    431534        1             5          4                 15   
5   3367565        1             6          2                  7   
6    550135        1             7          1                  9   
7   3108588        1             8          1                 14   
8   2295261        1             9          1                 16   
9   2550362        1            10          4                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   
5                    19.0         196                  1          1   
6                    20.0         196                  1          1   
7                    14.0         196                  2          1   
8                     0.0         196                  4          1   
9                    30.0         196                  1          1   

  product_name  aisle_id  department_id  prices    price_range_loc  \
0         Soda        77              7     9.0  Mid-range product   
1         Soda        77              7     9.0  Mid-range product   
2         Soda        77              7     9.0  Mid-range product   
3         Soda        77              7     9.0  Mid-range product   
4         Soda        77              7     9.0  Mid-range product   
5         Soda        77              7     9.0  Mid-range product   
6         Soda        77              7     9.0  Mid-range product   
7         Soda        77              7     9.0  Mid-range product   
8         Soda        77              7     9.0  Mid-range product   
9         Soda        77              7     9.0  Mid-range product   

   orders_day_of_the_week  Busiest_days busiest_period_of_day  
0                       2  Regular busy        Average orders  
1                       3  Slowest days        Average orders  
2                       3  Slowest days        Average orders  
3                       4  Slowest days        Average orders  
4                       4  Slowest days        Average orders  
5                       2  Regular busy        Average orders  
6                       1  Busiest days        Average orders  
7                       1  Busiest days        Average orders  
8                       1  Busiest days        Average orders  
9                       4  Slowest days        Average orders

In [8]:
ords_prods_merge.shape

(32404859, 17)

In [9]:
ords_prods_merge.describe()

order_id       user_id  order_number     order_dow  \
count  3.240486e+07  3.240486e+07  3.240486e+07  3.240486e+07   
mean   1.710745e+06  1.029372e+05  1.714230e+01  2.738867e+00   
std    9.872988e+05  5.946610e+04  1.753532e+01  2.090077e+00   
min    2.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.559470e+05  5.142200e+04  5.000000e+00  1.000000e+00   
50%    1.711049e+06  1.026160e+05  1.100000e+01  3.000000e+00   
75%    2.565499e+06  1.543890e+05  2.400000e+01  5.000000e+00   
max    3.421083e+06  2.062090e+05  9.900000e+01  6.000000e+00   

       order_hour_of_day  days_since_prior_order    product_id  \
count       3.240486e+07            3.032876e+07  3.240486e+07   
mean        1.342515e+01            1.110408e+01  2.559866e+04   
std         4.246380e+00            8.779064e+00  1.408400e+04   
min         0.000000e+00            0.000000e+00  1.000000e+00   
25%         1.000000e+01            5.000000e+00  1.354400e+04   
50%         1.300000e+01            8.000000e+00  2.530200e+04   
75%         1.600000e+01            1.500000e+01  3.794700e+04   
max         2.300000e+01            3.000000e+01  4.968800e+04   

       add_to_cart_order     reordered      aisle_id  department_id  \
count       3.240486e+07  3.240486e+07  3.240486e+07   3.240486e+07   
mean        8.352547e+00  5.895873e-01  7.119612e+01   9.919792e+00   
std         7.127071e+00  4.919087e-01  3.821139e+01   6.281485e+00   
min         1.000000e+00  0.000000e+00  1.000000e+00   1.000000e+00   
25%         3.000000e+00  0.000000e+00  3.100000e+01   4.000000e+00   
50%         6.000000e+00  1.000000e+00  8.300000e+01   9.000000e+00   
75%         1.100000e+01  1.000000e+00  1.070000e+02   1.600000e+01   
max         1.450000e+02  1.000000e+00  1.340000e+02   2.100000e+01   

             prices  orders_day_of_the_week  
count  3.240486e+07            3.240486e+07  
mean   7.790171e+00            2.738867e+00  
std    4.241978e+00            2.090077e+00  
min    1.000000e+00            0.000000e+00  
25%    4.200000e+00            1.000000e+00  
50%    7.400000e+00            3.000000e+00  
75%    1.130000e+01            5.000000e+00  
max    2.500000e+01            6.000000e+00

# Step 2 Aggregate 'mean' for the entire dataframe 

In [10]:
ords_prods_merge.groupby('product_name')

In [12]:
# Aggregate the 'mean' for the entire data

pd.options.display.max_rows = None

In [13]:
ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

# Step 3 Analyzing the results and how the results differ from the lesson's subset.  The first thing is the lesson was a subset or a portion of the data which was 1 million rows. Where in the task, this is the entire dataframe of 32,404,859 rows. In the lesson only eight 'department_id' came up, and in the task 21 'department_id' showed. Also the 'mean' in the lesson verses the task are different between the 'department_id'.  For example, the 'mean' for 'department_id' 4, in the lesson the mean is 18.825780 while in the task the 'mean' is 17.811403 for the same 'department_id'.  This difference is most likely caused by taking an aggregation of a subset verses the entire dataframe. 

# Step 4 Follow the exercise instructions for creating loyalty flag for existing customers using 'transform() and loc() functions.

In [14]:
# Aggregating data with 'transform()' function

ords_prods_merge['max_order'] = ords_prods_merge.groupby('user_id')['order_number'].transform(np.max)

In [15]:
ords_prods_merge.head(15)

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0    2539329        1             1          2                  8   
1    2398795        1             2          3                  7   
2     473747        1             3          3                 12   
3    2254736        1             4          4                  7   
4     431534        1             5          4                 15   
5    3367565        1             6          2                  7   
6     550135        1             7          1                  9   
7    3108588        1             8          1                 14   
8    2295261        1             9          1                 16   
9    2550362        1            10          4                  8   
10   2968173       15            15          1                  9   
11   1870022       15            17          2                 16   
12   1911383       15            18          2                 11   
13   2715276       15            21          1                  9   
14    487368       15            22          1                 10   

    days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                      NaN         196                  1          0   
1                     15.0         196                  1          1   
2                     21.0         196                  1          1   
3                     29.0         196                  1          1   
4                     28.0         196                  1          1   
5                     19.0         196                  1          1   
6                     20.0         196                  1          1   
7                     14.0         196                  2          1   
8                      0.0         196                  4          1   
9                     30.0         196                  1          1   
10                     7.0         196                  2          0   
11                     8.0         196                  6          1   
12                     7.0         196                  1          1   
13                     7.0         196                  1          1   
14                    14.0         196                  1          1   

   product_name  aisle_id  department_id  prices    price_range_loc  \
0          Soda        77              7     9.0  Mid-range product   
1          Soda        77              7     9.0  Mid-range product   
2          Soda        77              7     9.0  Mid-range product   
3          Soda        77              7     9.0  Mid-range product   
4          Soda        77              7     9.0  Mid-range product   
5          Soda        77              7     9.0  Mid-range product   
6          Soda        77              7     9.0  Mid-range product   
7          Soda        77              7     9.0  Mid-range product   
8          Soda        77              7     9.0  Mid-range product   
9          Soda        77              7     9.0  Mid-range product   
10         Soda        77              7     9.0  Mid-range product   
11         Soda        77              7     9.0  Mid-range product   
12         Soda        77              7     9.0  Mid-range product   
13         Soda        77              7     9.0  Mid-range product   
14         Soda        77              7     9.0  Mid-range product   

    orders_day_of_the_week  Busiest_days busiest_period_of_day  max_order  
0                        2  Regular busy        Average orders         10  
1                        3  Slowest days        Average orders         10  
2                        3  Slowest days        Average orders         10  
3                        4  Slowest days        Average orders         10  
4                        4  Slowest days        Average orders         10  
5                        2  Regular busy        Average orders         10  
6                        1  Busiest days        Average orders         10  
7

In [ ]:
# Deriving columns with 'loc()' function

In [16]:
ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [17]:
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [18]:
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [19]:
ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: loyalty_flag, dtype: int64

In [20]:
ords_prods_merge[['user_id', 'loyalty_flag', 'order_number']].head(60)

user_id      loyalty_flag  order_number
0         1      New customer             1
1         1      New customer             2
2         1      New customer             3
3         1      New customer             4
4         1      New customer             5
5         1      New customer             6
6         1      New customer             7
7         1      New customer             8
8         1      New customer             9
9         1      New customer            10
10       15  Regular customer            15
11       15  Regular customer            17
12       15  Regular customer            18
13       15  Regular customer            21
14       15  Regular customer            22
15       19      New customer             2
16       19      New customer             5
17       19      New customer             7
18       21  Regular customer            10
19       31  Regular customer            10
20       31  Regular customer            17
21       43  Regular customer             6
22       43  Regular customer             9
23       52  Regular customer             1
24       52  Regular customer             2
25       52  Regular customer             3
26       52  Regular customer             5
27       52  Regular customer             7
28       52  Regular customer             9
29       52  Regular customer            10
30       52  Regular customer            11
31       52  Regular customer            12
32       52  Regular customer            15
33       52  Regular customer            17
34       52  Regular customer            18
35       52  Regular customer            19
36       52  Regular customer            21
37       67  Regular customer             1
38       67  Regular customer             2
39       67  Regular customer             3
40       67  Regular customer             4
41       67  Regular customer             5
42       67  Regular customer             6
43       67  Regular customer             8
44       67  Regular customer             9
45       67  Regular customer            10
46       67  Regular customer            11
47       67  Regular customer            12
48       67  Regular customer            13
49       67  Regular customer            14
50       67  Regular customer            16
51       67  Regular customer            18
52       67  Regular customer            19
53       67  Regular customer            21
54       67  Regular customer            22
55       67  Regular customer            23
56       81      New customer             3
57       81      New customer             5
58       82  Regular customer             1
59       82  Regular customer             3

# Step 5 The marketing team wants to know if there is a difference between spending habits of the three types of customers (Loyal Customer, Regular Customer, and New Customer). Try to determine whether the prices of products purchased by loyal customers differ from the regular or new customers.

In [21]:
# Check if 'Loyal customers' purchases differ from 'Regular customers' and 'New customers'

ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max']})

prices           
                      mean  min   max
loyalty_flag                         
Loyal customer    7.772903  1.0  25.0
New customer      7.800248  1.0  25.0
Regular customer  7.797393  1.0  25.0

In [ ]:
# The spending average of the three groups is close. However, New customers purchase a little more with 7.800248 on average, 
# then Regular customers at 7.797393 on average and Loyalcustomers at 7.772903 on average. 
# The average price of products purchased does not seem to be a significant difference between the three groups.

# Step 6 The team wants to target different types of spenders in their marketing.  This can be achieved by looking at the prices of the items people are buying. Create a spending flag for each user based on the average price across all their orders using the following:

# If the mean of the prices of products purchased by a user is lower than 10, then flag them as a "Low spender"

# If the mean of the prices of products purchased by a user is higher than or equal to 10, then flag them as a "High spender"

In [36]:
# Aggregating data with 'transform()' function

ords_prods_merge['type_of_spender'] = ords_prods_merge.groupby('user_id')['prices'].transform(np.mean)

In [37]:
ords_prods_merge.head(15)

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0    2539329        1             1          2                  8   
1    2398795        1             2          3                  7   
2     473747        1             3          3                 12   
3    2254736        1             4          4                  7   
4     431534        1             5          4                 15   
5    3367565        1             6          2                  7   
6     550135        1             7          1                  9   
7    3108588        1             8          1                 14   
8    2295261        1             9          1                 16   
9    2550362        1            10          4                  8   
10   2968173       15            15          1                  9   
11   1870022       15            17          2                 16   
12   1911383       15            18          2                 11   
13   2715276       15            21          1                  9   
14    487368       15            22          1                 10   

    days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                      NaN         196                  1          0   
1                     15.0         196                  1          1   
2                     21.0         196                  1          1   
3                     29.0         196                  1          1   
4                     28.0         196                  1          1   
5                     19.0         196                  1          1   
6                     20.0         196                  1          1   
7                     14.0         196                  2          1   
8                      0.0         196                  4          1   
9                     30.0         196                  1          1   
10                     7.0         196                  2          0   
11                     8.0         196                  6          1   
12                     7.0         196                  1          1   
13                     7.0         196                  1          1   
14                    14.0         196                  1          1   

   product_name  ...    price_range_loc  orders_day_of_the_week  Busiest_days  \
0          Soda  ...  Mid-range product                       2  Regular busy   
1          Soda  ...  Mid-range product                       3  Slowest days   
2          Soda  ...  Mid-range product                       3  Slowest days   
3          Soda  ...  Mid-range product                       4  Slowest days   
4          Soda  ...  Mid-range product                       4  Slowest days   
5          Soda  ...  Mid-range product                       2  Regular busy   
6          Soda  ...  Mid-range product                       1  Busiest days   
7          Soda  ...  Mid-range product                       1  Busiest days   
8          Soda  ...  Mid-range product                       1  Busiest days   
9          Soda  ...  Mid-range product                       4  Slowest days   
10         Soda  ...  Mid-range product                       1  Busiest days   
11         Soda  ...  Mid-range product                       2  Regular busy   
12         Soda  ...  Mid-range product                       2  Regular busy   
13         Soda  ...  Mid-range product                       1  Busiest days   
14         Soda  ...  Mid-range product                       1  Busiest days   

   busiest_period_of_day  max_order      loyalty_flag type_of_spender  \
0         Average orders         10      New customer        6.367797   
1         Average orders         10      New customer        6.367797   
2         Average orders         10      New customer        6.367797   
3         Average orders         10      New customer        6.367797   
4         Average orders         10      New customer        6.367797   
5         A

In [ ]:
# Deriving columns with 'loc()' function

In [38]:
ords_prods_merge.loc[ords_prods_merge['type_of_spender'] >= 10, 'spending_flag'] = 'High spender'

In [39]:
ords_prods_merge.loc[ords_prods_merge['type_of_spender'] < 10, 'spending_flag'] = 'Low spender'

In [41]:
ords_prods_merge['spending_flag'].value_counts(dropna = False)

Low spender     32285165
High spender      119694
Name: spending_flag, dtype: int64

In [42]:
ords_prods_merge[['user_id', 'spending_flag', 'prices']].head(60)

user_id spending_flag  prices
0         1   Low spender     9.0
1         1   Low spender     9.0
2         1   Low spender     9.0
3         1   Low spender     9.0
4         1   Low spender     9.0
5         1   Low spender     9.0
6         1   Low spender     9.0
7         1   Low spender     9.0
8         1   Low spender     9.0
9         1   Low spender     9.0
10       15   Low spender     9.0
11       15   Low spender     9.0
12       15   Low spender     9.0
13       15   Low spender     9.0
14       15   Low spender     9.0
15       19   Low spender     9.0
16       19   Low spender     9.0
17       19   Low spender     9.0
18       21   Low spender     9.0
19       31   Low spender     9.0
20       31   Low spender     9.0
21       43   Low spender     9.0
22       43   Low spender     9.0
23       52   Low spender     9.0
24       52   Low spender     9.0
25       52   Low spender     9.0
26       52   Low spender     9.0
27       52   Low spender     9.0
28       52   Low spender     9.0
29       52   Low spender     9.0
30       52   Low spender     9.0
31       52   Low spender     9.0
32       52   Low spender     9.0
33       52   Low spender     9.0
34       52   Low spender     9.0
35       52   Low spender     9.0
36       52   Low spender     9.0
37       67   Low spender     9.0
38       67   Low spender     9.0
39       67   Low spender     9.0
40       67   Low spender     9.0
41       67   Low spender     9.0
42       67   Low spender     9.0
43       67   Low spender     9.0
44       67   Low spender     9.0
45       67   Low spender     9.0
46       67   Low spender     9.0
47       67   Low spender     9.0
48       67   Low spender     9.0
49       67   Low spender     9.0
50       67   Low spender     9.0
51       67   Low spender     9.0
52       67   Low spender     9.0
53       67   Low spender     9.0
54       67   Low spender     9.0
55       67   Low spender     9.0
56       81   Low spender     9.0
57       81   Low spender     9.0
58       82   Low spender     9.0
59       82   Low spender     9.0

# Step 7 To send relevant notifications to users within the app, if they want to buy the same item again, the team wants 'frequent' vs 'non-frequent' customers. Create an 'order_frequency_flag' that marks the regularity of a user's ordering behavior according to the median in the "days_since_prior_order" column. The criteria for the flag is as follows:

# If the median of "days_since_prior_order" is higher than 20 then the customer should be labeled a "Non-frequent customer"

# If the median is higher than 10 and lower than or equal to 20, then the customer should be labeled a "Regular customer"

# If the median is lower than or equal to 10, then the customer should be labeled a "Frequent customer"

In [15]:
# Aggregating data with 'transform()' function

ords_prods_merge['regularity_of_user'] = ords_prods_merge.groupby('user_id')['days_since_prior_order'].transform(np.median)

In [16]:
ords_prods_merge.head(15)

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0    2539329        1             1          2                  8   
1    2398795        1             2          3                  7   
2     473747        1             3          3                 12   
3    2254736        1             4          4                  7   
4     431534        1             5          4                 15   
5    3367565        1             6          2                  7   
6     550135        1             7          1                  9   
7    3108588        1             8          1                 14   
8    2295261        1             9          1                 16   
9    2550362        1            10          4                  8   
10   2968173       15            15          1                  9   
11   1870022       15            17          2                 16   
12   1911383       15            18          2                 11   
13   2715276       15            21          1                  9   
14    487368       15            22          1                 10   

    days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                      NaN         196                  1          0   
1                     15.0         196                  1          1   
2                     21.0         196                  1          1   
3                     29.0         196                  1          1   
4                     28.0         196                  1          1   
5                     19.0         196                  1          1   
6                     20.0         196                  1          1   
7                     14.0         196                  2          1   
8                      0.0         196                  4          1   
9                     30.0         196                  1          1   
10                     7.0         196                  2          0   
11                     8.0         196                  6          1   
12                     7.0         196                  1          1   
13                     7.0         196                  1          1   
14                    14.0         196                  1          1   

   product_name  aisle_id  department_id  prices    price_range_loc  \
0          Soda        77              7     9.0  Mid-range product   
1          Soda        77              7     9.0  Mid-range product   
2          Soda        77              7     9.0  Mid-range product   
3          Soda        77              7     9.0  Mid-range product   
4          Soda        77              7     9.0  Mid-range product   
5          Soda        77              7     9.0  Mid-range product   
6          Soda        77              7     9.0  Mid-range product   
7          Soda        77              7     9.0  Mid-range product   
8          Soda        77              7     9.0  Mid-range product   
9          Soda        77              7     9.0  Mid-range product   
10         Soda        77              7     9.0  Mid-range product   
11         Soda        77              7     9.0  Mid-range product   
12         Soda        77              7     9.0  Mid-range product   
13         Soda        77              7     9.0  Mid-range product   
14         Soda        77              7     9.0  Mid-range product   

    orders_day_of_the_week  Busiest_days busiest_period_of_day  \
0                        2  Regular busy        Average orders   
1                        3  Slowest days        Average orders   
2                        3  Slowest days        Average orders   
3                        4  Slowest days        Average orders   
4                        4  Slowest days        Average orders   
5                        2  Regular busy        Average orders   
6                        1  Busiest days        Average orders   
7                        1  Busiest days        Average orders   
8              

In [ ]:
# Deriving columns with 'loc()' function

In [17]:
ords_prods_merge.loc[ords_prods_merge['regularity_of_user'] > 20, 'frequency_flag'] = 'Non-frequent customer'

In [18]:
ords_prods_merge.loc[(ords_prods_merge['regularity_of_user'] <= 20) & (ords_prods_merge['regularity_of_user'] > 10), 'frequency_flag'] = 'Regular customer'

In [19]:
ords_prods_merge.loc[ords_prods_merge['regularity_of_user'] <= 10, 'frequency_flag'] = 'Frequent customer'

In [20]:
ords_prods_merge['frequency_flag'].value_counts(dropna = False)

Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636437
NaN                             5
Name: frequency_flag, dtype: int64

In [21]:
ords_prods_merge[['user_id', 'regularity_of_user', 'days_since_prior_order']].head(60)

user_id  regularity_of_user  days_since_prior_order
0         1                20.5                     NaN
1         1                20.5                    15.0
2         1                20.5                    21.0
3         1                20.5                    29.0
4         1                20.5                    28.0
5         1                20.5                    19.0
6         1                20.5                    20.0
7         1                20.5                    14.0
8         1                20.5                     0.0
9         1                20.5                    30.0
10       15                10.0                     7.0
11       15                10.0                     8.0
12       15                10.0                     7.0
13       15                10.0                     7.0
14       15                10.0                    14.0
15       19                 6.0                     6.0
16       19                 6.0                     8.0
17       19                 6.0                     6.0
18       21                 7.5                     6.0
19       31                 6.0                     7.0
20       31                 6.0                     8.0
21       43                10.0                     4.0
22       43                10.0                     3.0
23       52                 8.0                     NaN
24       52                 8.0                     7.0
25       52                 8.0                     7.0
26       52                 8.0                     8.0
27       52                 8.0                     8.0
28       52                 8.0                     7.0
29       52                 8.0                     8.0
30       52                 8.0                     7.0
31       52                 8.0                     6.0
32       52                 8.0                     8.0
33       52                 8.0                     6.0
34       52                 8.0                    14.0
35       52                 8.0                     8.0
36       52                 8.0                    30.0
37       67                 7.0                     NaN
38       67                 7.0                     5.0
39       67                 7.0                     7.0
40       67                 7.0                     8.0
41       67                 7.0                     8.0
42       67                 7.0                     7.0
43       67                 7.0                     7.0
44       67                 7.0                     7.0
45       67                 7.0                     6.0
46       67                 7.0                     7.0
47       67                 7.0                     8.0
48       67                 7.0                     6.0
49       67                 7.0                     7.0
50       67                 7.0                     9.0
51       67                 7.0                     4.0
52       67                 7.0                     9.0
53       67                 7.0                     5.0
54       67                 7.0                    14.0
55       67                 7.0                    14.0
56       81                30.0                    14.0
57       81                30.0                    30.0
58       82                11.5                     NaN
59       82                11.5                     5.0

# Step 8 Ensure notebook is clean and structured and codes are well documented.

# Step 9 Export dataframe as a pickle file and store it correctly in "Prepared Data" folder

In [23]:
ords_prods_merge.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged.pkl'))